File: src/ray.rs
use crate::vec3::Vec3;

pub type Point3 = Vec3;

pub struct Ray {
    orig: Point3,
    dir: Vec3,
}

impl Ray {
    pub fn new(origin: Point3, direction: Vec3) -> Ray {
        Ray {
            orig: origin,
            dir: direction,
        }
    }

    pub fn origin(&self) -> Point3 {
        self.orig
    }

    pub fn direction(&self) -> Vec3 {
        self.dir
    }

    pub fn at(&self, t: f64) -> Point3 {
        self.orig + t * self.dir
    }
}

pub struct HitRecord<'material> {
    pub t: f64,
    pub point: Vec3,
    pub normal: Vec3,
    pub front_face: bool,
    pub material: &'material Material,
    pub u: f64,
    pub v: f64,
}

pub trait Hittable {
    fn hit(&self, ray: &Ray, t_min: f64, t_max: f64) -> Option<HitRecord>;
}

File: ray.go
package main

// Ray represents a line in 3D space defined by an origin and a direction.
type Ray struct {
	Origin    Point3
	Direction Vec3
}

// At returns the point along the ray at parameter t: P(t) = A + t*b
func (r Ray) At(t float64) Point3 {
	// P(t) = Origin + Direction * t
	return r.Origin.Add(r.Direction.MultiplyScalar(t))
}
